### PR TITLE
Allow type-safe calls to be used in InOrder

### LINKED ISSUE #70: type-safe calls can't be used in gomock.InOrder
**Actual behavior** When using type-safe mode (`-typed`), the calls returned from `EXPECT().Method()` are typed. This is great, and a feature I've been waiting for for a long time. However, it makes it impossible to use them in `gomock.InOrder`, since `InOrder` expects a `*gomock.Call`.

**Expected behavior** I expected `gomock.InOrder` to work seamlessly.

**To Reproduce** Steps to reproduce the behavior

1.  Generate mocks using `-typed`
2. Use `gomock.InOrder` with calls of that mock

**Additional Information**

-   gomock mode (reflect or source): reflect
-   gomock version or git ref: v0.2.0
-   golang version: Go 1.21

**Triage Notes for the Maintainers**

Failed CI run on quic-go after generating typed mocks: https://github.com/quic-go/quic-go/actions/runs/5988902088/job/16244553880?pr=4051 (PR: https://github.com/quic-go/quic-go/pull/4051)

### PR DESCRIPTION
**Pull request by EstebanOlmedo**
This modifies `InOrder` to receive variadic `any` instead of `*Call`, allowing users to use this function with type-safe generated Calls.

This implementation uses a type assertion to check if the provided arguments are `*Call`s or reflection to get the `*Call` when the arguments wrap one (generated code). If neither of the two cases are fullfiled then `InOrder` panics.

Fix #70.

### CODE REVIEW

**Comment by marten-seemann** [comment]
> If neither of the two cases are fullfiled then that argument is ignored.

Would it make sense to panic? Today it's not possible to pass an invalid call to `InOrder`. It would be nice if I'd be notified right away if I mess up `InOrder`, otherwise this will be very hard to debug.

File: gomock/call.go [hunk 1]
```diff
@@ -435,12 +435,40 @@ func (c *Call) call() []func([]any) []any {
...
 // InOrder declares that the given calls should occur in order.
-func InOrder(calls ...*Call) {
+// Ignores every argument that isn't a *Call or wraps one.
+func InOrder(args ...any) {
+        var calls []*Call
+        for i := 0; i < len(args); i++ {
+                if call := getCall(args[i]); call != nil {
+                        calls = append(calls, call)
+                }
+        }
```
Comment (sywhang) [inline gomock/call.go:445]: we should panic here. Silently passing through this isn't going to help -- clearly indicate the expected parameter type should be either a `*Call` type or a generated mock type which should already embed `*Call`.

**Comment by marten-seemann** [comment]
Alternatively, what about a type-safe(r) way of doing this: Having `InOrder` accept an `interface { Call() *Call }`?

File: gomock/call.go [hunk 2]
```diff
@@ -435,12 +435,43 @@ func (c *Call) call() []func([]any) []any {
...
+// It panics if the type of any of the arguments isn't *Call or a generated
+// mock with an embedded *Call.
+func InOrder(args ...any) {
+	var calls []*Call
+	for i := 0; i < len(args); i++ {
+		if call := getCall(args[i]); call != nil {
+			calls = append(calls, call)
+			continue
+		}
+		panic("InOrder parameters type must be either a *Call or a generated mock with an embedded *Call")
```
Comment (r-hang) [inline gomock/call.go:447]: Can we also print as part of the error message additional details on what was passed through that caused the error?

**Comment by marten-seemann** [comment]
Tested in quic-go in https://github.com/quic-go/quic-go/pull/4057. Seems to work. I also tested that it panic if I pass things other than gomock calls to `InOrder`.

🚢 it!

**Review by marten-seemann — APPROVED** [review]

**Review by sywhang — APPROVED** [review]

**Comment by nkostoulas** [comment]
This breaks call that pass a `[]*Call` slice 😞 .

**Pull request merged**